### PR TITLE
Polynomials over commutative rings added

### PR DESCRIPTION
**Pull request by AkermanRydbeck**
Hello,

Here is an implementation of polynomials over commutative rings.

Cheers,
Carl

### CODE REVIEW

File: Cubical/Algebra/CommRing/Instances/Polynomials.agda [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+open import Cubical.Foundations.Prelude
+open import Cubical.Algebra.CommRing
+open import Cubical.Algebra.Polynomials
+
+private
+  variable
+    ℓ : Level
+
+
+ℙ : (R : CommRing ℓ) → CommRing ℓ
```
Comment (mortberg) [inline Cubical/Algebra/CommRing/Instances/Polynomials.agda:14]: Bättre namn, typ `Poly`

Comment (mortberg) [reply Cubical/Algebra/CommRing/Instances/Polynomials.agda:14]: Typen kan vara ` CommRing ℓ → CommRing ℓ`

File: Cubical/Algebra/CommRing/Instances/Polynomials.agda [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ℙ R = (PolyMod.Poly R) , str
+    where
+      open CommRingStr (snd R)
+      str : CommRingStr (PolyMod.Poly R)
+      CommRingStr.0r str = PolyMod.0P R
```
Comment (mortberg) [inline Cubical/Algebra/CommRing/Instances/Polynomials.agda:19]: Testa att ta bort ` CommRingStr.`, funkar det inte ändra till `open CommRingStr` ovan (så inget `(snd R)`)

File: Cubical/Algebra/Polynomials.agda [hunk 3]
```diff
@@ -0,0 +1,784 @@
...
+    A : Type ℓ
+
+module PolyMod (R' : CommRing ℓ) where
+  private
+    R = fst R'
+  open CommRingStr (snd R') public
+
+-------------------------------------------------------------------------------------------
+-- First definition of a polynomial.
+-- A polynomial a₁ +  a₂x + ... + aⱼxʲ of order j is represented as a list [a₁, a₂, ...,aⱼ]
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials.agda:40]: order <- degree

och ett polynom är en lista "modulo trailing zeros"

File: Cubical/Algebra/Polynomials.agda [hunk 4]
```diff
@@ -0,0 +1,784 @@
...
+                  (BProp : {p : Poly} → isProp (B p)) where
+   f : (p : Poly) → B p
+   f = Elim.f B []* cons* (toPathP (BProp (transport (λ i → B (drop0 i)) (cons* 0r [] []*)) []*))
+
+
+  module Rec (B : Type ℓ')
+             ([]* : B)
+             (cons* : R → B → B)
+             (drop0* : cons* 0r []* ≡ []*)
+                          (Bset : isSet B) where
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials.agda:77]: Indentering

File: Cubical/Algebra/Polynomials.agda [hunk 5]
```diff
@@ -0,0 +1,784 @@
...
+    f (a ∷ p) = cons* a (f p)
+    f (drop0 i) = drop0* i
+
+
+
+--------------------------------------------------------------------------------------------------
+-- Second definition of a polynomial. The purpose of this second definition is to
+-- facilitate the proof that the first definition is a set. The two definitions are
+-- then shown to be equivalent.
+-- A polynomial a₀ +  a₁x + ... + aⱼxʲ of order j is represented as a function f
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials.agda:94]: degree

File: Cubical/Algebra/Polynomials.agda [hunk 6]
```diff
@@ -0,0 +1,784 @@
...
+-- such that for i ∈ ℕ we have  f(i) = aᵢ if i ≤ j, and 0 for i > j
+--------------------------------------------------------------------------------------------------
+
+  PolyFun : Type ℓ
+  PolyFun = Σ[ p ∈ (ℕ → R) ] (∃[ n ∈ ℕ ] ((m : ℕ) → n ≤ m → p m ≡ 0r))
+
+
+  IsSetPolyFun : isSet PolyFun
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials.agda:102]: `isSetPolyFun`

File: Cubical/Algebra/Polynomials.agda [hunk 7]
```diff
@@ -0,0 +1,784 @@
...
+  IsSetPolyFun = isSetΣSndProp (isSetΠ (λ x → isSetCommRing R')) λ f x y → squash x y
+
+
+  IsZero : ℕ → Bool
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials.agda:106]: Flytta till `Cubical.Data.Nat.Base` och byt namn till `isZero`   (om det blir konflikter pga import av Bool där så får vi ser vad vi gör)

File: Cubical/Algebra/Polynomials.agda [hunk 8]
```diff
@@ -0,0 +1,784 @@
...
+  PolyFun→Poly+ : (q : PolyFun) → Σ[ p ∈ Poly ] Poly→Fun p ≡ q .fst
+  PolyFun→Poly+ (f , pf) = rec lem (λ x → help f (x .fst) (x .snd) ,
+                                               funExt (bar f (fst x) (snd x))
+                                   ) pf
+    where
+    lem : isProp (Σ[ p ∈ Poly ] Poly→Fun p ≡ f)
+    lem (p , α) (p' , α') =
+      ΣPathP (polyEq p p' (α ∙ sym α'), isProp→PathP (λ i → (isSetΠ λ _ → is-set) _ _) _ _)
+
+    help : (p : ℕ → R) (n : ℕ) → ((m : ℕ) → n ≤ m → p m ≡ 0r) → Poly
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials.agda:207]: Byt namn till `rem1`

File: Cubical/Algebra/Polynomials.agda [hunk 9]
```diff
@@ -0,0 +1,784 @@
...
+    help p zero h = []
+    help p (suc n) h = p 0 ∷ help (λ x → p (suc x)) n (λ m x → h (suc m) (suc-≤-suc x))
+
+    bar : (f : ℕ → R) (n : ℕ) → (h : (m : ℕ) → n ≤ m → f m ≡ 0r) (m : ℕ) →
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials.agda:211]: Byt namn till `rem2`

File: Cubical/Algebra/Polynomials.agda [hunk 10]
```diff
@@ -0,0 +1,784 @@
...
+                                                                 Poly→Fun (help f n h) m ≡ f m
+    bar f zero h m = sym (h m zero-≤)
+    bar f (suc n) h zero = refl
+    bar f (suc n) h (suc m) = bar (λ x → f (suc x)) n (λ k p → h (suc k) (suc-≤-suc p)) m
+
+  PolyFun→Poly : PolyFun → Poly
+  PolyFun→Poly q = PolyFun→Poly+ q .fst
+
+  lemmaisSet : (p : Poly) → PolyFun→Poly (Poly→PolyFun p) ≡ p
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials.agda:220]: Byt namn till `PolyFun→Poly→PolyFun`

File: Cubical/Algebra/Polynomials.agda [hunk 11]
```diff
@@ -0,0 +1,784 @@
...
+  lemmaisSet p = polyEq _ _ (PolyFun→Poly+ (Poly→PolyFun p) .snd)
+
+
+
+--End of code by Mörtberg and Cavallo
+-------------------------------------
+
+  IsSetPoly : isSet Poly
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials.agda:228]: Byt namn `isSetPoly`

File: Cubical/Algebra/Polynomials.agda [hunk 12]
```diff
@@ -0,0 +1,784 @@
...
+  IsSetPoly = isSetRetract Poly→PolyFun
+                           PolyFun→Poly
+                           (λ p → lemmaisSet p)
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials.agda:231]: Behövs inget lambda här, utan `lemmaIsSet` räcker (eller `PolyFun→Poly→PolyFun` när du bytt namn då)

File: Cubical/Algebra/Polynomials.agda [hunk 13]
```diff
@@ -0,0 +1,784 @@
...
+  open RingTheory (CommRing→Ring R')
+  open GroupTheory (Ring→Group (CommRing→Ring R'))
+
+
+  pattern [_] x = x ∷ []
+
+
+  -- The proof that Poly is a set will be used in the propositional eliminations
+  -- for the last case in ElimProp
+  PolyIsSet : isSet Poly
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials.agda:249]: Kan tas bort, använd `isSetPoly` överallt istället

File: Cubical/Algebra/Polynomials.agda [hunk 14]
```diff
@@ -0,0 +1,784 @@
...
+---------------------------------------
+-- Definition
+-- Identity for addition of polynomials
+---------------------------------------
+  0P : Poly
+  0P = []
+
+
+  --ReplicatePoly(n,p) returns 0 ∷ 0 ∷ ... ∷ [] (n zeros)
+  ReplicatePoly : (n : ℕ)  → Poly
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials.agda:262]: Byt namn till `replicatePoly0`

File: Cubical/Algebra/Polynomials.agda [hunk 15]
```diff
@@ -0,0 +1,784 @@
...
+  ReplicatePoly zero  = 0P
+  ReplicatePoly (suc n) = 0r ∷ ReplicatePoly n
+
+
+  --The empty polynomial has multiple equal representations on the form 0 + 0x + 0 x² + ...
+  ReplicatedZeros=[] : ∀ (n : ℕ) → ReplicatePoly n ≡ 0P
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials.agda:268]: `replicatePoly0Is0P`

File: Cubical/Algebra/Polynomials.agda [hunk 16]
```diff
@@ -0,0 +1,784 @@
...
+-- Definition
+-- subtraction of polynomials
+-----------------------------
+  Poly- : Poly → Poly
+  Poly- [] = []
+  Poly- (a ∷ p) = (- a) ∷ (Poly- p)
+  Poly- (drop0 i) = (cong (_∷ []) (inv1g) ∙ drop0) i
+
+  -- Double negation (of subtraction of polynomials) is the identity mapping
+  Poly-Poly-P=P : (p : Poly) → Poly- (Poly- p) ≡ p
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials.agda:283]: `Poly-Poly-`

File: Cubical/Algebra/Polynomials.agda [hunk 17]
```diff
@@ -0,0 +1,784 @@
...
+  Poly-Poly-P=P = ElimProp.f (λ x → Poly- (Poly- x) ≡ x)
+                             (refl)
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials.agda:285]: parenteser behövs inte runt refl

File: Cubical/Algebra/Polynomials.agda [hunk 18]
```diff
@@ -0,0 +1,784 @@
...
+  Poly+Lid =  ElimProp.f (λ p → ([] Poly+ p ≡ p) )
+                         (refl)
+                         (λ r p prf → refl)
+                         (λ x y → PolyIsSet _ _ x y)
+
+
+
+  -- [] is the right identity for Poly+
+  Poly+Rid : ∀ p → (p Poly+ [] ≡ p)
+  Poly+Rid = λ p → refl
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials.agda:322]: `Poly+Rid p = refl`

File: Cubical/Algebra/Polynomials.agda [hunk 19]
```diff
@@ -0,0 +1,784 @@
...
+                                                     sym (Poly*RDistrPoly+ (a PolyConst* q)
+                                                                           (0r ∷ (p Poly* q)) r)
+                                                                           ⟩
+                 ((((a ∷ p) Poly* q) Poly* r)) ∎)
+               (λ x y i q r  → PolyIsSet _ _ (x q r) (y q r) i)
+
+
+
+-----------------------------------
+-- Main Theorem                  --
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials.agda:776]: Ta bort `PolyCommRing` här och referera till `Cubical.Algebra.CommRing.Instances.Polynomials` i en kommentar

**Review by mortberg — CHANGES_REQUESTED** [review]
Tack så mycket!

**Review by mortberg — APPROVED** [review]

**Comment by mortberg** [comment]
Oops, there is a clash for `isZero`: https://github.com/agda/cubical/runs/5465312974?check_suite_focus=true

**Comment by felixwellen** [comment]
And I am really happy we have this definition of polynomials now! And also that it is a set - I didn't know that it works this nicely.

**Comment by mortberg** [comment]
> You can just hide `isZero` on import in `HornerForms.agda`, then everything type checks.

Yes, that should be the best solution. @AkermanRydbeck  Can you do it please?

**Comment by mortberg** [comment]
> And I am really happy we have this definition of polynomials now! And also that it is a set - I didn't know that it works this nicely.

Yes, I'm also very pleased that we managed to prove that it's a set without truncating. It's very nice to not have to deal with the truncation and to be able to eliminate into arbitrary types.

**Comment by AkermanRydbeck** [comment]
Hello, glad you like it Felix. 

isZero is hidden now.

/Carl

**Pull request merged**